---
title: "Comparison with previous qPCR findings"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE)
library(knitr)
library(kableExtra)
library(tidyverse)
```


## Introduction

Existing qPCR findings for HLA expression and its association with regulatory SNPs and diseases provide an important reference for any study on HLA expression. Many researchers in the field consider qPCR as a golden standard for HLA expression. Here, because of the indisputable importance of placing our novel findings in the context of well established functional studies, we provide a comparison with previous qPCR studies. Of course, this comparison has limitations: when comparing across studies we often are dealing with different individuals, cell types, and techniques. Thus, our efforts to compare our RNA-seq findings to those of published papers using qPCR are at best a first approximation.

In our review of the qPCR literature on HLA expression, we found it difficult to compare results across studies because qPCR data are usually not comparable among different HLA loci, due to the different experimental techniques which are developed for each locus.

Regarding allele-level estimates, they are not obtained directly with qPCR. Although some efforts have been made to develop allele-specific primers (Pan et al, 2018), to our knowledge, these have not yet been validated and used to obtain expression levels. Allele-based expression levels from qPCR (e.g., Kulkarni et al, 2011; Ramsuram et al, 2015; Ramsuram et al, 2017) are usually imputed from the locus-level expression levels (in the graphs included in those papers, locus-level expression is plotted twice, one point for each allele). Attempts have been made to improve the imputation with the use of a linear model of expression~genotype (e.g., Ramsuram et al, 2015). Besides the technical and biological differences between studies, the imputed nature of allele-level estimates from qPCR makes us not not surprised that many differences will emerge when we compare our RNA-seq data with qPCR data at the HLA-allele level.




## Methods

### Data

Expression estimates for *HLA-A*, *HLA-B* and *HLA-C* as reported by [Ramsuram et al, HMG, 2015](https://doi.org/10.1093/hmg/ddv158), [Ramsuram et al, JI, 2017](http://www.jimmunol.org/content/198/6/2320), and [Kulkarni et al, PNAS, 2013](https://doi.org/10.1073/pnas.1312237110), were kindly provided by corresponding authors.


### Comparison of expression levels between qPCR and RNAseq

Because our study and the previously published papers with qPCR expression data involved different tissues and samples, they cannot be directly compared (e.g., using a correlation analysis). Instead, we chose to ask whether the the ordering of HLA lineages given their expression levels differs among studies.  

Because this ordering is affected by the sample sizes of each lineage, the variation of individuals for the same lineage, and the degree of differentiation among lineages, we performed a comparison of our data (HLApers on GEUVADIS) with those in qPCR-based studies only for pairs of lineages with $N \ge 10$, and with statistically different distributions of expression values according to a pairwise Mann-Whitney U test with FDR correction for multiple testing, as available in the R statistical environment.


### Association of expression with SNPs and diseases

Existing qPCR findings for the association of SNPs and HLA expression provide an important reference for any study on HLA expression.

In our Table S3, we assessed the degree of independence of the eQTLs we mapped in respect to previously reported regulatory SNPs, including those from qPCR studies, but there we showed results for the best association for each eQTL only. Here we use that same analysis, and dedicate more space to a comparison with previous SNPs with experimental support (Kulkarni et al, 2011; Petersdorf et al, 2015; Vince et al, 2016; Raj et al, 2016; Ou et al, 2018). We evaluate independence in terms of D', $r^2$ and RTC score.




## Results

### Comparison of lineage ordering across studies

A first impression of the relative ordering of lineages according to the expression levels suggests substantial differences (Figure 1).

\hfill\break
  
```{r, fig.align='center'}
include_graphics("./compare_qpcr/FigSA.png")
```

**Figure 1.** Lineage-level expression in previous qPCR studies and RNA-seq (HLApers) for *HLA-A*, *HLA-B* and *HLA-C*. We included only lineages present in at least 10 individuals in both GEUVADIS RNA-seq data and qPCR data. Y-axis: Transcript per Million for RNA-seq, and $2^{-\Delta \Delta Ct}$ for qPCR.

\hfill\break

However, the variation among individuals for the same lineage is often very high, and the sample sizes vary among lineages, so that the ordering of lineage-level expression based on mean or median values may be within the expected range of variation for a single lineage. We thus restricted our analyses to pairs of lineages with $N \ge 10$, and with statistically different distributions of expression values according to a pairwise Mann-Whitney U test with FDR correction for multiple testing (in both qPCR and RNA-seq).

Overall for *HLA-A*, *HLA-B* and *HLA-C* we found 33 instances where the expression was significantly different for a pair of lineages in both qPCR and RNA-seq datasets (Table 1). 

\hfill\break

**Table 1** Significant lineage pairs in both qPCR and RNA-seq in a pairwise Mann-Whitney test.

\begin{center}

```{r}
read_tsv("./compare_qpcr/results_significant.tsv") %>%
    select(Lineage1 = a1, Lineage2 = a2, `pvalue (pcr)` = 3, `pvalue (hlapers)` = 4) %>%
    kable(digits = 5)
```

\end{center}

\hfill\break

Of the significant pairs, for 21/33 cases the results were concordant (i.e. we found the same ordering of expression in our study and in qPCR-based studies). However, the results were markedly different among loci: for *HLA-C*, 13 out of 13 significant pairs were concordant, for *HLA-B*, 4 out 6 had the same ordering. On the other hand, for *HLA-A*, qPCR and RNA-seq results were concordant in only 4 out 14 pairs (Table 2).


\hfill\break

**Table 2.** Number of lineage pairs concordant between qPCR and RNA-seq, total of significant pairs tested, and percentage of concordance for *HLA-A*, *HLA-B* and *HLA-C*.

\begin{center}

```{r}
read_tsv("./compare_qpcr/summary_results.tsv") %>% 
    mutate(locus = paste0("HLA-", locus),
           `Percentage of concordance` = same_dir/n_significant * 100) %>%
    rename(Locus = 1, `Concordant with qPCR` = 2, `Total significant pairs tested` = 3) %>%
    kable(digits = 1)
``` 

\end{center}

\hfill\break

The low number of significant pairs for *HLA-B* can be explained by the low differentiation of expression levels among *HLA-B* lineages in the qPCR data (7 pairs out of 55 were significant, whereas RNA-seq had 31), and by the low sample sizes of some lineages in qPCR. For example, B\*13 and B\*51 are the most and the least expressed lineages in both qPCR and RNA-seq, but in the qPCR data there are only 11 individuals with B\*13 and 14 individuals with B\*51, which lowers the power of the statistical test. Furthermore, the variability is so high that some individuals with the least expressed lineage (B\*51) have higher expression than some individuals with B\*13.



### Independence regarding previous regulatory SNPs with validation support.

We find that the eQTLs we mapped for HLA-C is very highly correlated at the haplotypic level with variants described by Kulkarni et al (2011) and Vince et al (2016) ($0.97 \le D’ \le 1$), which were shown to regulate HLA-C expression via a miRNA binding site at the 3’UTR, and at an Oct1 binding site at the promoter region, respectively. The minor allele frequencies of our variants are lower than those in the previous studies, and the variation at our SNPs is nested within that of the previously validated SNPs, i.e. the allele present in our SNPs predict the allele in the previous SNP, but the opposite is not true. The RTC test suggest that accounting for the previously validated SNP does not erase the signal represented by our SNPs, which means that, although there are strong haplotypic association, our SNPs seem to be driving the biological signal in our sample. This is also the case for the rank 1 eQTL which we mapped for HLA-DQB1 and a variant reported by Raj et al (2016) as part of a super-enhancer which regulates HLA-DQA1, HLA-DQB1 and HLA-DRB1. For HLA-DP we also found strong association, both in terms of D’ and $r^2$, between our SNPs and those previously validated for HLA-DPB1 (Petersdorf et al, 2015) and HLA-DPA1 (Ou et al, 2018). However, the RTC test suggest that, although the allele in one SNP predicts the allele in previously validated SNP, accounting for the previous SNP does not remove the signal of the SNP we mapped (Table 2).

\hfill\break

**Table 3.** Relationship between our eQTLs and previous SNPs with validation support ("Valid-SNP"). r2: r-squared; AF: allele frequency; p: the backward pass p-value from QTLtools for the validated SNP (NA when SNP did not pass to the backward stage).  

```{r}
"./geuvadis_reanalysis/eqtl_mapping/transcriptomemapping/hla_personalized/rtc/top_previous_qtls/validated_variants/validated_variants_results.csv" %>%
    read_csv() %>%
    filter(d_prime > 0.65) %>%
    select(Gene = 1, rank = 2, eQTL = 3, `Valid-SNP` = 4, Study = 5,  
           `r2` = 6, `D'` = 7, `AF (eQTL)` = 8, `AF (Valid-SNP)` = 9,
           RTC = 10, p = 11) %>%
    mutate(Study = recode(Study, kulkarni2011 = "Kulkarni 2011", 
                          vince2017 = "Vince 2016", 
                          xl9_raj2016 = "Raj 2016",
                          ou2018 = "Ou 2018", 
                          petersdorf2015 = "Petersdorf 2015")) %>%
    kable(digits = 2) %>%
    kable_styling(latex_options = "scale_down")
```

\hfill\break

Therefore, this exercise of comparing our findings on HLA regulatory SNPs with variants with some previous validation, we find some agreements and disagreements. We note however that it is plausible that previously validated variants will not have a strong effect in our sample, because different tissues have different mechanisms for regulating a gene, so a mechanism identified in a previous study may not be among the most relevant processes involved in the regulation of HLA-C in our sample/tissue. Furthermore, previous studies have not queried all the variation in the region, but only a few candidate SNPs. 


## References


[Kulkarni et al. Differential microRNA regulation of HLA-C expression and its association with HIV control. Nature. 2011;472(7344).](https://doi.org/10.1038/nature09914)

[Ou G et al. Relationship between HLA-DPA1 mRNA expression and susceptibility to hepatitis B. Journal of Viral Hepatitis. 2019;26.](https://doi.org/10.1111/jvh.13012)

[Petersdorf E et al. High HLA-DP Expression and Graft-versus-Host Disease. The New England Journal of Medicine. 2015;373(7).](https://doi.org/10.1056/NEJMoa1500140)

[Pan N et al. Quantification of classical HLA class I mRNA by allele-specific,real-time polymerase chain reaction for most Han individuals. HLA. 2019;91.](https://doi.org/10.1111/tan.13186)

[Raj P et al. Regulatory polymorphisms modulate the expression of HLA class II molecules and promote autoimmunity. eLife. 2016;5.](https://doi.org/10.7554/eLife.12089)

[Ramsuram V et al. Epigenetic regulation of differential HLA-A allelic expression levels. Human Molecular Genetics. 2015;24(15).](https://doi.org/10.1093/hmg/ddv158)

[Ramsuram V et al. Sequence and Phylogenetic Analysis of the Untranslated Promoter Regions for HLA Class I Genes. The Journal of Immunology. 2017;198.](https://doi.org/10.4049/jimmunol.1601679)

[Vince N et al. HLA-C Level Is Regulated by a Polymorphic Oct1 Binding Site in the HLA-C Promoter Region. The American Journal of Human Genetics. 2016;99(6).](https://doi.org/10.1016/j.ajhg.2016.09.023)

