---
title: Expression estimation and eQTL mapping of HLA genes
author: 
  - Meyer Lab$^1$
  - Dermitzakis Lab$^2$
institute: 
  - $^1$University of SÃ£o Paulo
  - $^2$University of Geneva
output: 
  beamer_presentation:
  mathjax: "http://example.com/mathjax/MathJax.js?config=TeX-AMS-MML_HTMLorMML"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, comment='', message=FALSE, warning=FALSE)
```

## The importance of HLA expression phenotypes

- Association to cancer phenotypes
- Association to HIV/AIDS progression
- Transplantation

## Importance of developing HLA expression pipeline

- Many datasets of genome-wide data are available, but HLA estimates may be inaccurate
- Better eQTL mapping
- HLA allele-level estimates
- Understand how the regulatory architecture is related to HLA lineages

## Strategy

\includegraphics[width=\linewidth]{../../hla_manuscript/figures/method.png}

\*We explored approach using 3 tools: GEM, STAR (standard aligners) kallisto (kmer-based). 

## Simulation: Indices convering HLA diversity recover more reads

Typing accuracy: 100%

\includegraphics[width=\linewidth]{../simulation/plots/star_prop_mapped.png}

<!--
- statement I would make here: confirms expectation that single reference leads
  to loss of reads, and that this is not randomly distributed but concentrated
  on alleles which are more divergent from reference sequence.

- To think about: even for HLA-B, most alleles have more than 75% reads mapped.
  Consistent with the similarity in expression estimates among indices. 
-->

## Quality assessment

| Gene    | not aligned  | aligned to different ref | false positive reads | 
|:-------:|:------------:|:------------------------:|:--------------------:|
|HLA-A    |         0    |                     1.66 | 0                    |
|HLA-B    |         0    |                     0.2  | 0                    |
|HLA-C    |         0    |                     0    | 0                    |
|HLA-DQA1 |         0    |                     0.4  | 0                    |
|HLA-DQB1 |         0    |                     0    | 0                    |
|HLA-DRB1 |         0    |                     0.4  | 0                    |

## Geuvadis Data

**Typing accuracy**

| Gene    | accuracy  |
|:-------:|:---------:|
|HLA-A    |  97%      |
|HLA-B    |  98%      |
|HLA-C    |  96%      |
|HLA-DQB1 |  95%      |
|HLA-DRB1 |  98%      |

\*Gold standard:
1000 Genomes individuals HLA-typed by Gourraud et al. (2014). PLoS ONE 9(7): e97282.

## Aligner choice does not influence expression estimate

\includegraphics[width=\linewidth]{../geuvadis_reanalysis/expression/plots/kallisto_vs_star_TPM.png}

## Validation 1: Correlation of expression

\includegraphics[width=.8\linewidth]{./hlacorrelations.png}

## Validation 2: Class II genes and their transactivator

\includegraphics[width=\linewidth]{../geuvadis_reanalysis/expression/plots/trans_activ_corrs.png}

## Correlation with Geuvadis published estimates

\includegraphics[width=\linewidth]{./star_vs_geuvadis_final.png}

## Taking the uncorrected Geuvadis data and running myself the correction 

\includegraphics[width=\linewidth]{./star_vs_geuvadis.png}

<!-- 
Conclusion: resonable locus-level for some loci. Very important to understand
why the new quantification does a poorer job. In particular, at some level, the
new quantification should be similar to our quantification using the standard
reference genome, which in turn is similar to our IMGT index run. So it seems
that some detail of the alignment influences the outcome. 
-->

## New quantifications 

\includegraphics[width=\linewidth]{./star_vs_geuvadis_new.png}

## Mapping to REF chromosomes may be enough for gene-level estimates

\includegraphics[width=.9\linewidth]{./star_imgt_vs_chr.png}

# eQTLs

## Number of eGenes

\includegraphics[width=\linewidth]{./eGenes_number.png}

## eQTLs at HLA genes

\includegraphics[width=.7\linewidth]{../geuvadis_reanalysis/qtls/plots/qtls_landscape_star.png}

## RTC with Geuvadis eQTLs

\fontsize{7pt}{6}\selectfont

| Gene    | rank | variant   | geuvadis gene    | geuvadis variant | RTC |
|:-------:|:----:|:---------:|:----------------:|:----------------:|:---:|
|HLA-C    |   0  |rs41561715 | HLA-C            | rs115899777      | 0.95|
|HLA-DQA1 |   0  |rs373884000| HLA-DRB1         | rs116405062      | 0.97|
|HLA-DQA1 |   1  |rs9274722  | HLA-DQA1/HLA-DQB1| rs9274660        | 0.92|
|HLA-DQB1 |   0  |rs1770     | HLA-DQA1/HLA-DQB1| rs9274660        | 0.98|
|HLA-DRB1 |   0  |rs9270698  | HLA-DQA1/HLA-DQB1| rs9274660        | 0.92|
|HLA-DRB1 |   2  |rs145607970| HLA-DQA1/HLA-DQB1| rs9274660        | 0.93|

## How allele-level expression is estimated in the HLA field

\includegraphics[width=\linewidth]{./ramsuran_hlaA_lineages.png}

Ramsuram et al. Human Molecular Genetics, 2015, 24 (15).

## Relationship between eQTLs and HLA alleles

\includegraphics[width=.75\linewidth]{../geuvadis_reanalysis/qtls/plots/lineage_and_effects.png}

## Expression of alleles from different genes on the same vs on different haplotypes

\includegraphics[width=\linewidth]{../geuvadis_reanalysis/expression/plots/b_vs_c.png}

\includegraphics[width=\linewidth]{../geuvadis_reanalysis/expression/plots/dqa_vs_drb.png}

## Conclusions

- A pipeline which accounts for the HLA variation provides better expression estimates, especially for HLA-DQA1 and HLA-DRB1;

- HLA alleles are associated with disease phenotypes, and the pipeline allows quantification of allele-level expression;

- Do the eQTLs explain variation in HLA allele expression?

- Much of the HLA literature is focused on the association of disease phenotypes with HLA haplotypes, but we see no correlation of expression attributed to haplotype membership.
