---
title: "Report"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, comment = '')
```

```{r pkgs}
library(knitr)
library(tidyverse)
```

# Contrasting qPCR and RNAseq quantification of HLA class I expression 

## NCI-USP collaboration

## Document prepared by Vitor RC Aguiar and Diogo Meyer, May 26 2017

The goal of this study was to compare two technologies used to estimate HLA
expression through the quantification of mRNA. We performed an RNAseq assay on
a set of 96 individuals for which qPCR expression estimates were available, and
for a subset of which HLA-C antibody-based HLA cell surface expression was
available.

The RNAseq assay was based on our two-step strategy, where we first virtually
genotype each individual (using the full list of IMGT alleles) and then perform
a second mapping where we only include the two alleles the individual carries in
the aligner's index. This provides allele level expression estimates, which can
be converted to locus level my summing the expression of both alleles.

## Typing concordance with NCI genotyping

We first quantfied our ability to correctly genotype individuals based on the
RNAseq data. Our method is quite simple, and in very general terms simply
involves identifying which of the alleles in the IMGT database have more NGS
reads aligning to it. The table below identifies the proportion of genotypes for
which our RNAseq-based genotyping matched that reported by conventional HLA
typing.

\*Concordance: the proportion of the called alleles that are concordant with the NCI typing

\*Allele calls were compared at the maximum resolution possible at each pair

```{r}
read_tsv("./expression/genotyping_accuracies.tsv") %>%
  filter(th == 0.20) %>%
  select(locus, concordance = accuracy) %>%
  kable()
```

## Library Sizes

```{r}
include_graphics('./plots/library_sizes.png')
```

## Correlation between qPCR and RNAseq expression estimates

We then compared the expression level estimates between RNAseq and qPCR. There
is no correlation for HLA-B, some for HLA-A, and the highest is seen for HLA-C.

```{r}
include_graphics('./plots/seq_vs_pcr.png')
```

## HLA-C expression: RNA vs Antibody

Both qPCR and RNAseq strategies have similarly low correlations with the
antibody-base expression estimates.

```{r}
include_graphics('./plots/ab_vs_rna.png')
```

# RNAseq: TPM vs B2M-normalized counts

Sample 66K00755 has an abnormally low B2M expression

```{r}
include_graphics('./plots/tpm_vs_B2MnormCounts.png')
```

# RNAseq: TPM vs GAPDH-normalized counts

```{r}
include_graphics('./plots/tpm_vs_GAPDHnormCounts.png')
```

# RNAseq: TPM vs ACTB-normalized counts

```{r}
include_graphics('./plots/tpm_vs_ACTBnormCounts.png')
```

# Coverage

The coverage measure is the gene-level coverage, which is the sum of the coverage, at each position, over the 2 alleles of a locus (black line in the plot below).

```{r}
include_graphics('./plots/covs_measure.png')
```

It does not make sense to compare coverage with a normalized measure of expression, such as qPCR and TPM. For example, two samples with the same coverage at a given gene may have different TPM values, because TPM normalization depends on the mRNA composition of each sample.

On the other hand, coverage is very well correlated with kallisto's estimated counts.

```{r}
include_graphics('./plots/covs_expression.png')
```

## RNAseq: expression estimates per lineage

Our approach also allows assigning expression level estimates to allelic
lineages. We compare the lineage-level expression for RNAseq (first set of
panels) and qPCR (second set of panels, using imputed expression by regression analysis as done in the NCI group papers).

## RNAseq

```{r}
include_graphics('./plots/rnaseq_lineages.png')
```

## qPCR

```{r}
include_graphics('./plots/nci_lineages.png')
```




