---
title: "Report"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, comment = "")
```

```{r pkgs, echo = FALSE}
library(knitr)
library(tidyverse)
```

# eQTLs

**All analyses were carried out using European individuals only**

## PCA of genotypes

```{r}
include_graphics("./plots/genotype_pca.png")
```

## Number of eGenes according to method of phenotype correction

```{r}
include_graphics("./plots/pca_vs_peer.png")
```

## Distribution of eQTLs around the TSS

### kallisto

```{r}
include_graphics("./plots/qtls_landscape_kallisto.png")
```

### STAR

```{r}
include_graphics("./plots/qtls_landscape_star.png")
```

## HLA lineages

```{r}
include_graphics("./plots/lineage_and_effects.png")
```

# Comparison with Geuvadis

## RTC

Some of our eQTLs seem to mark the same biological signal as the eQTLs found by
Geuvadis (RTC > 0.9)

```{r}
"./qtls_star/rtc/geuvadis_eqtls/results.tsv" %>%
  read_tsv() %>%
  kable()
```

## slope and p-value

```{r}
geuvadis_qtls <- 
  read_tsv("./qtls_star/rtc/geuvadis_eqtls/catalog.tsv",
	   col_names = c("variant", "info")) %>%
  select(-variant) %>%
  separate(info, c("geuvadis_variant", "geuvadis_gene", "geuvadis_slope",
		   "geuvadis_pvalue"), sep = ":")

star_qtls <- read_tsv("./plots/eqtls.tsv") %>%
  select(gene = phen_id, rank, variant = var_id, slope, pvalue = nom_pval) %>%
  mutate_at(vars(slope, pvalue), function(x) round(x, digits = 2))

pval_df <-
  star_qtls %>%
  filter(rank == 0L) %>%
  left_join(geuvadis_qtls, by = c("gene" = "geuvadis_gene")) %>%
  select(gene, variant, geuvadis_variant, slope, geuvadis_slope, 
	 pvalue, geuvadis_pvalue)

kable(pval_df)
```

## LD between our best SNP (STAR) and Geuvadis best SNP

```{r}
read_tsv("./qtls_star/rtc/geuvadis_eqtls/ld_star_geuvadis_bestSNPs.tsv") %>%
  unite("Geuvadis_STAR", c("allele_geuvadis", "allele_star"), sep = "_") %>%
  rename(f_allele_geuv = f_geuvadis, f_allele_star = f_star) %>%
  kable(digits = 2)
```

# Association to GWAS traits

```{r}
read_tsv("./qtls_star/rtc/gwas/results.tsv") %>%
  select(gene, rank, variant, gwas_variant, rtc, trait, studies = link) %>%
  group_by(gene, rank, variant, gwas_variant, rtc) %>%
  summarize(trait = paste(trait, collapse = "/"), 
	    studies = paste(studies, collapse = " ")) %>%
  kable(digits = 2)
```

# Trans-eQTLs

- Approximate pass as described on QTLtools website

```{r}
read_tsv("./qtls_star/trans/trans_results.tsv") %>%
  kable()
```
