---
title: "Report"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, comment = "")
```

```{r pkgs, echo = FALSE}
library(knitr)
library(tidyverse)
```

# eQTLs

**All analyses were carried out using European individuals only**

## PCA of genotypes

```{r}
include_graphics("./plots/genotype_pca.png")
```

## Number of eGenes according to method of phenotype correction

```{r}
include_graphics("./plots/pca_vs_peer.png")
```

## Distribution of eQTLs around the TSS

### kallisto

```{r}
include_graphics("./plots/qtls_landscape_kallisto.png")
```

### STAR

```{r}
include_graphics("./plots/qtls_landscape_star.png")
```

## HLA lineages

```{r}
include_graphics("./plots/lineage_and_effects.png")
```

# RTC

## Geuvadis

Some of our eQTLs seem to mark the same biological signal as the eQTLs found by
Geuvadis (RTC > 0.9)

```{r}
"./qtls_star/rtc/geuvadis_eqtls/results.tsv" %>%
  read_tsv() %>%
  kable()
```

## Association with GWAS traits

```{r}
read_tsv("./qtls_star/rtc/gwas/results.tsv") %>%
  select(gene, rank, variant, gwas_variant, rtc, trait, studies = link) %>%
  kable()
```

## Trans-eQTLs

- Approximate pass as described on QTLtools website

```{r}
read_tsv("./qtls_star/trans/trans_results.tsv") %>%
  kable()
```
