---
title: "Report"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, comment = '')
```

```{r pkgs}
library(tidyverse)
library(knitr)
```

# kallisto

## Genotyping

```{r}
read_tsv("./expression/kallisto/genotyping_accuracies.tsv") %>%
  filter(th == .05) %>%
  mutate(accuracy = accuracy * 100) %>%
  select(locus, `accuracy (%)` = accuracy) %>%
  kable()
```

## Expression

```{r}
include_graphics('./plots/hlasimul.png')
```

# STAR + Salmon

## Genotyping

```{r}
read_tsv("./expression/star/genotyping_accuracies.tsv") %>%
  filter(th == .05) %>%
  mutate(accuracy = accuracy * 100) %>%
  select(locus, `accuracy (%)` = accuracy) %>%
  kable()
```

## Expression

```{r}
include_graphics("./plots/star_simul.png")
```

# Quality assessment 

Percentage of simulated reads not aligned:

```{r}
read_tsv("./reads_not_aligned_hla.tsv") %>% kable()
```

Percentage of simulated reads from each HLA gene that aligned to a different reference:

```{r}
read_tsv("./reads_false_neg_hla.tsv") %>% kable()
```

Percentage of simulated reads gained by each HLA gene (reads simulated from other references)

```{r}
read_tsv("./reads_false_pos_hla.tsv") %>% kable()
```
