---
title: "Report"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, comment = '')
```

```{r pkgs}
library(tidyverse)
library(knitr)
```

# Notes:

Index nomenclature:

- pri: Primary assembly of the reference genome, containing the reference chromosomes and scaffolds (no alternate haplotypes)

- imgt: "pri" index supplemented with IMGT references


# Genotyping

```{r}
read_tsv("./PEreads_75bp/expression/star/imgt/genotyping_accuracies_1.tsv") %>%
  filter(th == .05) %>%
  mutate(accuracy = accuracy * 100) %>%
  select(locus, `accuracy (%)` = accuracy) %>%
  kable()
```

# Expression

```{r}
include_graphics("./plots/star_prop_mapped.png")
```

# Quality assessment 

## Percentage of simulated reads not aligned, aligned exclusively to a different reference, or aligned to original reference but as a multimap

Here we can see that, for some genes, there is a high percentage of reads which
are not aligned, or aligned to a gene different from the one the read was
simulated from.

```{r}
sample_ids <- sprintf("sample_%02d", 1:50)

read_summary_imgt <- 
    file.path("./PEreads_75bp/expression/star/imgt/mappings_2",
	      sample_ids, "read_summary_hla.tsv") %>%
    setNames(sample_ids) %>%
    map_df(read_tsv, .id = "subject") %>%
    group_by(gene_from) %>%
    summarize_at(vars(perc_not_aligned:perc_aligned_to_original_uniquely), mean) %>%
    ungroup()

read_summary_pri <- 
    file.path("./PEreads_75bp/expression/star/pri/mappings",
	      sample_ids, "read_summary_hla.tsv") %>%
    setNames(sample_ids) %>%
    map_df(read_tsv, .id = "subject") %>%
    group_by(gene_from) %>%
    summarize_at(vars(perc_not_aligned:perc_aligned_to_original_uniquely), mean) %>%
    ungroup()

list("HLA-personalized" = read_summary_imgt, 
     "Reference" = read_summary_pri) %>%
bind_rows(.id = "index") %>%
arrange(gene_from, index) %>%
kable(digits = 2)
```

To further understand the read "loss" or "gain" we examined, for each HLA gene,
the percentage of alignments stratified by mapped or source gene, respectively.

So, the size of each point in the plots below represents the percentage of
alignments involving that gene pair.


### Percentage of aligments involving reads simulated from HLA in x axis but mapping to other gene

```{r}
include_graphics("./plots/alignments_to_diff_gene.png")
```

### Percentage of alignments involving reads not simulated from HLA in x axis but still mapping to them

```{r}
include_graphics("./plots/alignments_from_diff_gene.png")
```


# Comparisons between indices and aligners

## kallisto vs STAR-Salmon; HLA-diversity index

```{r}
include_graphics("./plots/kallisto_vs_star_TPM.png")
```

## kallisto vs STAR-Salmon; Reference chromosomes only

```{r}
include_graphics("./plots/kallisto_vs_star_PRI_TPM.png")
```

## kallisto; HLA-diversity vs Reference chromosomes only

```{r}
include_graphics("./plots/kallisto_imgt_vs_PRI_TPM.png")
```

## STAR-Salmon; HLA-diversity vs Reference chromosomes only

```{r}
include_graphics("./plots/star_imgt_vs_PRI_TPM.png")
```

# Isoforms

The simulation above is idealized in the sense that reads are generated from the
coding sequences and mapped back to them. However, real transcription occurs via
mRNA transcripts, containing UTRs and different combinations of exons.

In order to evaluate the performance of the HLA-supplemented index to map the
isoforms, we simulated reads from the hypothetical individual homozygote for the
reference HLA allele at the 7 HLA loci. Ground truth isoform counts were based
on the estimated counts for the individual NA20508 quantified with the reference
transcriptome index (protein coding transcripts only).

We see a reduction in expression in the TPM estimate for HLA-DQA1. Including the
UTRs or trimming the index to 750bp do not solve the problem.

```{r}
x <- read_tsv("./PEreads_75bp/isoforms/results.tsv")

kable(x, digits = 3)
```
