---
title: "Report on typing errors"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, comment = '')
```

```{r pkgs}
devtools::load_all("~/hlaseqlib/")
library(tidyverse)

get_diff_pos <- function(index, allele.x, allele.y) {
    
    regex.y <- sub("\\*", "\\\\*", allele.y) %>% 
        paste0("^", ., "(:|[NQLS]|$)")
    
    m_err <- index[names(index) == allele.x] %>%
        as.character() %>%
        strsplit("")
    
    m_true <- index[grepl(regex.y, names(index))][1] %>%
        as.character() %>%
        strsplit("")
    
    m <- do.call(rbind, c(m_err, m_true))

    diffs <- apply(m, 2, function(x) length(unique(x)) > 1) %>%
        which()
    
    if (length(diffs) == 0L) diffs <- NA

    tibble(allele.x = allele.x,
           allele.y = names(m_true),
           POS = diffs)
}

plot_cov <- function(sample_id) {
    
    df <- filter(plot_df, subject == sample_id)
    df_vline <- filter(alleles_df, subject == sample_id)
    
    ggplot(data = df, aes(x = POS, y = DEPTH, color = call, group = call)) +
        geom_vline(data = df_vline, aes(xintercept = POS), color = "green") +
        geom_line(size = 3, alpha = 0.5) +
        ggrepel::geom_label_repel(aes(label = label), na.rm = TRUE, show.legend = FALSE) +
        ggsci::scale_color_aaas() +
        facet_wrap(~locus, ncol = 2) +
        theme_bw() +
        labs(color = "allele")
}
```

```{r data}
index <- 
    Biostrings::readDNAStringSet("~/hlaexpression/imgt_index_v2/imgt_index_witDots.fa")

samples <- readLines("./samples_with_errs.txt")

typing_err <- read_tsv("./typing_errors.tsv") %>%
    rename(kgp_id = subject) %>%
    left_join(select(geuvadis_info, ena_id, name), by = c("kgp_id" = "name")) %>%
    select(subject = ena_id, kgp_id, locus, allele.x, allele.y, correct) %>%
    arrange(subject, locus) %>%
    mutate(allele.y = sub("^([^/]+).*$", "\\1", allele.y),
           allele.y = hla_trimnames(allele.y, 3)) %>%
    mutate_at(vars(allele.x, allele.y), ~paste0("IMGT_", .))

coverage <- paste0("./mappings_1/", samples, "_imgt.cov") %>%
    setNames(samples) %>%
    map_df(~read_tsv(., col_names = FALSE), 
           .id = "subject") %>%
    rename(allele = X1, POS = X2, DEPTH = X3)

alleles_df <- typing_err %>% 
    filter(subject %in% samples) %>%
    group_by(subject, locus) %>%
    mutate(h = 1:n()) %>%
    ungroup() %>%
    mutate(d = map2(allele.x, allele.y, get_diff_pos, index = index)) %>%
    select(subject, kgp_id, locus, h, d) %>%
    unnest(d) %>%
    unite(locus, c("locus", "h")) %>%
    select(subject, kgp_id, locus, goldStd = allele.y, called = allele.x, POS) %>%
    gather(call, allele, goldStd, called) %>%
    left_join(coverage, by = c("subject", "allele", "POS")) %>%
    select(subject, kgp_id, locus, call, allele, POS, DEPTH) %>%
    arrange(subject, locus, call)

plot_df <- distinct(alleles_df, subject, kgp_id, locus, call, allele) %>%
    inner_join(coverage, by = c("subject", "allele")) %>%
    mutate(allele = sub("IMGT_", "", allele)) %>%
    group_by(subject, locus, call) %>%
    mutate(label = ifelse(DEPTH == max(DEPTH), allele, NA)) %>%
    group_by(subject, locus, call, label) %>%
    mutate(label2 = ifelse(POS == max(POS), label, NA)) %>%
    ungroup() %>%
    select(subject, kgp_id, locus, call, allele, POS, DEPTH, label = label2)

err_df <- typing_err %>%
    distinct(subject, locus) %>%
    left_join(geuvadis_info, by = c("subject" = "ena_id")) %>%
    select(subject = name, locus) %>%
    mutate(locus = paste0("HLA-", locus))

kgp_called <- "../phase_hla_alleles/1000G_comparison/hla_diffs_to_1000Ghaps.tsv" %>%
    read_tsv() %>%
    inner_join(err_df, by = c("subject", "locus")) %>%
    rename(kgp_id = subject) %>%
    mutate(subject = convert_kgp_to_ena(kgp_id)) %>%
    select(subject, kgp_id, everything())

kgp_gold <- 
    "../phase_hla_alleles/1000G_comparison_fixedGenos/hla_diffs_to_1000Ghaps.tsv" %>%
    read_tsv() %>%
    inner_join(err_df, by = c("subject", "locus")) %>%
    rename(kgp_id = subject) %>%
    mutate(subject = convert_kgp_to_ena(kgp_id)) %>%
    select(subject, kgp_id, everything())

quant_called <- "./quantifications_2/processed_imgt_quants.tsv" %>%
    read_tsv() %>%
    left_join(geuvadis_info, by = c("subject" = "ena_id")) %>%
    select(subject, kgp_id = name, locus, allele, est_counts) %>%
    inner_join(err_df, by = c("kgp_id" = "subject", "locus")) %>%
    mutate(allele = gsub("IMGT_", "", allele))

quant_gold <- "../supplemented_fix_genos/quantifications_2/processed_imgt_quants.tsv" %>%
    read_tsv() %>%
    left_join(geuvadis_info, by = c("subject" = "ena_id")) %>%
    select(subject, kgp_id = name, locus, allele, est_counts) %>%
    inner_join(err_df, by = c("kgp_id" = "subject", "locus")) %>%
    mutate(allele = gsub("IMGT_", "", allele))
```


## HG00273

#### Mismatch at A\*11
```{r}
plot_cov(samples[1])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[1], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[1])
```

###### Gold
```{r}
kgp_gold %>% filter(subject == samples[1])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[1])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[1])
```


## NA12005

#### Mismatch at B\*27
```{r}
plot_cov(samples[2])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[2], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[2])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[2])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[2])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[2])
```


## NA11892

#### Mismatch at DQB1\*02
```{r}
plot_cov(samples[3])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[3], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[3])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[3])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[3])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[3])
```


### HG00256

#### Mismatch at DQB1\*02
```{r}
plot_cov(samples[4])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[4], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[4])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[4])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[4])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[4])
```


## NA10847

#### Mismatch at C\*07
```{r}
plot_cov(samples[5])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[5], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[5])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[5])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[5])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[5])
```


## NA20534

#### Mismatch at A\*11
```{r}
plot_cov(samples[6])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[6], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[6])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[6])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[6])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[6])
```


## HG00311

## Mismatch at A\*11
```{r}
plot_cov(samples[7])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[7], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[7])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[7])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[7])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[7])
```


## HG00261

#### False heterozygote DQB1\*05 in our calls
```{r}
plot_cov(samples[8])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[8], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[8])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[8])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[8])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[8])
```


## NA12287

### Mismatch at B\*51
```{r}
plot_cov(samples[9])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[9], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[9])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[9])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[9])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[9])
```


## NA11918

#### Mismatch at A\*02
```{r}
plot_cov(samples[10])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[10], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[10])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[10])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[10])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[10])
```


## NA07357

#### Mismatch at C\*07
```{r}
plot_cov(samples[11])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[11], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[11])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[11])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[11])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[11])
```


## NA12234

#### Mismatch at DQB1\*02
```{r}
plot_cov(samples[12])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[12], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[12])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[12])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[12])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[12])
```


## HG00115

#### Mismatch at A\*11
```{r}
plot_cov(samples[13])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[13], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[13])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[13])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[13])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[13])
```


## HG00171

#### Mismatch at C\*06
```{r}
plot_cov(samples[14])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[14], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[14])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[14])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[14])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[14])
```


## NA20585

#### Mismatch at A\*24 and C\*07
```{r}
plot_cov(samples[15])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[15], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[15])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[15])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[15])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[15])
```


## NA20538

#### Mismatch at C\*12
```{r}
plot_cov(samples[16])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[16], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[16])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[16])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[16])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[16])
```


## NA06986

#### Mismatch at DQB1\*02
```{r}
plot_cov(samples[17])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[17], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[17])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[17])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[17])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[17])
```


## NA11832

#### Mismatch at B\*27
```{r}
plot_cov(samples[18])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[18], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[18])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[18])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[18])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[18])
```


## NA11930

#### Mismatch at A\*02, C\*06 and DQB1\*02
```{r}
plot_cov(samples[19])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[19], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[19])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[19])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[19])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[19])
```


## NA20520

#### Mismatch at C\*05
```{r}
plot_cov(samples[20])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[20], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[20])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[20])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[20])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[20])
```


## NA11840

#### Mismatch at B\*27
```{r}
plot_cov(samples[21])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[21], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[21])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[21])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[21])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[21])
```


## HG00246

#### Mismatch at A\*11
```{r}
plot_cov(samples[22])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[22], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[22])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[22])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[22])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[22])
```


## NA20758

#### False A\*01 homozygote
```{r}
plot_cov(samples[23])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[23], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[23])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[23])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[23])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[23])
```


## NA20805

#### Mismatch C\*05
```{r}
plot_cov(samples[24])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[24], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[24])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[24])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[24])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[24])
```


## NA11994

#### Mismatch at A\*11 and C\*15
```{r}
plot_cov(samples[25])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[25], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[25])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[25])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[25])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[25])
```


## HG00380

#### Mismatch B\*42-B\*07
```{r}
plot_cov(samples[26])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[26], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[26])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[26])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[26])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[26])
```


## HG00109

#### We call a false A\*02 homozygote
```{r}
plot_cov(samples[27])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[27], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[27])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[27])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[27])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[27])
```


## NA20586

#### Mismatch at A\*11
```{r}
plot_cov(samples[28])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[28], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[28])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[28])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[28])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[28])
```


## NA11920

#### Mismatch at DRB1\*14
```{r}
plot_cov(samples[29])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[29], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[29])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[29])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[29])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[29])
```


## HG00360

#### We call a false A\*02 homozygote, also mismatch at B\*39
```{r}
plot_cov(samples[30])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[30], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[30])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[30])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[30])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[30])
```


## NA12045

#### Mismatch at DRB1\*14
```{r}
plot_cov(samples[31])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[31], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[31])
```

##### Gold

```{r}
kgp_gold %>% filter(subject == samples[31])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[31])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[31])
```


## NA12156

#### Mismatch at C\*06 and DQB1\*02
```{r}
plot_cov(samples[32])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[32], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[32])
```

###### Gold
```{r}
kgp_gold %>% filter(subject == samples[32])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[32])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[32])
```


## NA12058

#### Gold standard is homozygote for B and C, but we call heterozygote
```{r}
plot_cov(samples[33])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[33], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[33])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[33])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[33])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[33])
```


## HG00102

#### Mismatch at all alleles at A, B and C
```{r}
plot_cov(samples[34])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[34], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[34])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[34])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[34])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[34])
```


## HG00327

#### We call a false C\*05 heterozygote
```{r}
plot_cov(samples[35])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[35], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[35])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[35])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[35])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[35])
```


## NA11830

#### Mismatch at DQB1\*02
```{r}
plot_cov(samples[36])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[36], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[36])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[36])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[36])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[36])
```


## NA12272

#### Mismatch at A\*02, false heterozygote in our call for B, and mismatch at DRB1\*14
```{r}
plot_cov(samples[37])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[37], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[37])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[37])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[37])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[37])
```


## NA12286

#### Mismatch at DQB1\*02
```{r}
plot_cov(samples[38])
```

#### Depth at mismatch positions
```{r}
alleles_df %>% filter(subject == samples[38], !is.na(POS))
```

#### Mismatches to 1000G haplotypes

##### Called
```{r}
kgp_called %>% filter(subject == samples[38])
```

##### Gold
```{r}
kgp_gold %>% filter(subject == samples[38])
```

#### Final quantifications with called or gold standard genotypes:

##### Called
```{r}
quant_called %>% filter(subject == samples[38])
```

##### Gold
```{r}
quant_gold %>% filter(subject == samples[38])
```
